src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,755 @@
+/**
+ * @file hdbscan_impl.hpp
+ * Implementation of HDBSCAN.
+ */
+#ifndef __MLPACK_METHODS_HDBSCAN_HDBSCAN_IMPL_HPP
+#define __MLPACK_METHODS_HDBSCAN_HDBSCAN_IMPL_HPP
+
+#include "hdbscan.hpp"
+
+namespace mlpack {
+namespace hdbscan {
+
+/**
+ * Construct the HDBSCAN object with the given parameters.
+ */
+template<typename NeighborSearch ,
+         typename MetricType ,
+          template<typename TreeMetricType,
+                   typename TreeStatType,
+                   typename TreeMatType> class TreeType >
+HDBSCAN<NeighborSearch, MetricType, TreeType>::HDBSCAN(
+    const size_t minPoints,
+    bool allowSingleCluster) :
+    minPoints(minPoints),
+    allowSingleCluster(allowSingleCluster)
+{
+  // Nothing to do.
+}
+
+/**
+ * Performs HDBSCAN clustering on the data,returning the list of cluster assignments.
+ * The assigned clusters have values from 0 onwards. The noise points are labelled
+ * as (total nuber of points + 1)
+ *
+ * Dcore -> distance of k-th nearest neighbor
+ *          here k is minPoints
+ * 1. Find the dcore of all the points.
+ * 2. Then append dcore to end of the input data (dataWithDcore)
+ * 3. Compute MST using the Dual Tree Borvuka Method.
+ * 4. Sort the obtained MST on basis of length of edges
+ *    in ascending order.
+ * 5. Covert MST to single likage tree .
+ *    Single linkage tree is a clustering method in which
+ *    we go on clustering the points until a single cluster
+ *    is formed.
+ * 6. Condense the obtained single linkage tree
+ *    to form condensed tree. Condensed tree contains only those
+ *    clusters as parent which have more then minPoints number of
+ *    points.
+ * 7. From this condense tree obtain labels of all the points.
+ */
+template<typename NeighborSearch ,
+         typename MetricType ,
+          template<typename TreeMetricType,
+                   typename TreeStatType,
+                   typename TreeMatType> class TreeType >
+                   template<typename MatType>
+void HDBSCAN<NeighborSearch,  MetricType, TreeType>::Cluster(const MatType& data,
+                 arma::Row<size_t>& assignments)
+{
+  //genearting a matrix with the values of dcore
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighborSearch.Train(data);
+  neighborSearch.Search(minPoints, neighbors, distances);
+  arma::mat dataWithDcore = arma::conv_to<arma::mat>::from(data);
+  dataWithDcore.resize(dataWithDcore.n_rows+1, dataWithDcore.n_cols); 
+  dataWithDcore.row(dataWithDcore.n_rows-1) = distances.row(distances.n_rows-1);
+ 
+  //computing the mst of the given data and sort the obtained result
+  emst::DualTreeBoruvka<metric::HdbscanDistance, arma::mat, TreeType> dtb(dataWithDcore, true, metric);
```

Comment:
I think it satisfies the triangle inequality because : 
`hdbscanDistance(A, B) = max(dcore(A), dcore(B), euclideanDistance(A, B))`
`hdbscanDistance(A, B) >= euclideanDistance(A, B)`

Lets consider triangle ABC and side AB : 
if `hdbscanDistance(A, B) = euclideanDistance(A, B)`
then `hdbscanDistance(A, C) + hdbscanDistance(B, C) >=`
`euclideanDistance(A, C) + euclideanDistance(B,C) >= `
`euclideanDistance(A, B) = hdbscanDistance(A, B)`
other wise `hdbscanDistance(A, B) = dcore(A) or dcore(B).`
Lets say `hdbscanDistance(A, B) = dcore(A) `
then `hdbscanDistance(A, C) + hdbscanDistance(B, C) >=`
`hdbscanDistance(B, C) + dcore(A) >= dcore(A) = hdbscanDistance(A, B)`
We can argue similarly if `hdbscanDistance(A, B) = dcore(B) `
Hope, I am not mistaken.